doc/design/upgrade.md

```diff
@@ -0,0 +1,118 @@
+
+#### Refactor current VCH configuration and container VM configuration structure
+To make data migration easier, we'll need to refactor current VCH and container VM configuration structure.
+
+VirtualContainerHostConfigSpec is too big, which includes everything in one structure. We're not be able to update part of configuration structure, it will help to update part of data with separated structures based on functionality, e.g. network, storage, and common attributes.
+
+For ContainerVM configuration, there are a few attributes not used for container VM setup, e.g. in ExecutorConfig, Key, LayerID, RepoName, are all not container VM configuration related. Move out irrelated attributes can help the structure stability in the future. 
+
+### Transactional Operation/Roll Back
+VM snapshot is one good option for upgrade roll back. Before upgrade, vic-machine will create one snapshot of VCH. If anything wrong happens, vic-machine can switch back to the pre-upgrade snapshot.
+
+As snapshot does not change iso files, vic-machine still need to switch the appliance iso files back to old one.
+
+Note: VM snapshot does not persist data in serial port, as we will use datastore files through serial port for vch logs, after roll back, log files will have all error message during upgrade if VCH appliance VM is ever started with new configuration.
+
+### Limitation
+This document does not include approach for significant change which might break backward compatibility.
+- if datastore structure change break current data consumption model, new version cannot read back images from old version's data, we'll not try to migrate docker data during upgrade.
+- serial port is replaced by VMCI
+
+# Plan
+
+## Prerequisites
+
+- Need to have versioning support before we claim we can upgrade from this version.
+- Need to have VCH restart ready.
+- As all version is built into binary, we need to automatically generate new build after tag release.
+
+## Simple Upgrade
+
+At this point, we'd like to start from a simple solution, which does not include data migration. 
+
+Here is the upgrade workflow
+- Find existing VCH appliance
+- Query out VCH configuration
+- Verify new version and existing VCH's version
+- Valiate VCH configuration with new version's validator. If anything incompatible found, stop the upgrade. As no data migration yet.
+- Snapshot VCH appliance VM. - Should quiesce VCH during upgrade here (how?)
+- Poweroff VCH appliance VM if it's not
+- Upload iso files
+- Generate new VCH configuration from old one
+- Update VCH configuration version
+- Update container bootstrap iso file path
+- Update container VM configuration for anything required to support VCH restart(e.g. vch ip change, vch serial port number change, etc.)
+- Attach new appliance iso to VCH appliance VM
+- Power on VCH, and wait VCH initialization
+- Anything wrong in the above steps, switch appliance iso file back to old one, and then roll back to upgrade snapshot
```

Comment:
I think the following ordering will reduce our downtime:
 +- Find existing VCH appliance
 +- Query VCH configuration
 +- Verify new version and existing VCH's version
 +- Generate new VCH configuration from old one
 +- Valiate VCH configuration with new version's validator. If anything incompatible found, stop the upgrade. As no data migration yet.
 +- Upload iso files
 +- Prepare combined applianceVM configuration, both hardware spec (e.g. ISO path) and the extraconfig portion (e.g. VCH version)

--- everything until this point can be done without interrupting ongoing operation ---

 +- Snapshot VCH appliance VM - I don't think we need to quiesce it, we _could_ snapshot memory but that's only useful in the very short term for failure case, not rollback
 +- Poweroff VCH appliance VM
 +- Update container VM configuration with new config
 +- Power on VCH, and wait VCH initialization
 +- Anything wrong in the above steps, roll back to upgrade snapshot (this includes the ISO path so no need to explicitly revert that).
 +- Ensure appliance is powered on after rollback to snapshot.